### PR TITLE
Snow 1038314 use secure abstraction layer in base code

### PR DESCRIPTION
**Pull request by sfc-gh-pczajka**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [ ] I've added or updated integration tests to verify correctness of my new code.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
Change all found file operations to SecurePath operations (and update secure_path accordingly)

Usages I found:
* `*open*`
* `pathlib.X`: `write_text`, `read_text`, `touch`, `mkdir`
* `os.makedirs`
* `shutil.*`
* `tempfile.*` - one usage left in snowflake/cli/plugins/snowpark/venv.py - reason: may require big changes in SecurePath.temporary_directory implementation, will be moved into another ticket

SecurePath updates:
* add support for `pathlib.Path.write_text`
* add support for `shutil.move`
* add support for `shutil.copytree(dirs_exits_ok=True)`

### CODE REVIEW

File: src/snowflake/cli/api/secure_path.py [hunk 1]
```diff
@@ -125,44 +139,80 @@ def open(  # noqa: A003
...
-            encoding=encoding,
-            errors=errors,
-            newline=newline,
-        ) as fd:
+        with self._path.open(mode=mode, **open_kwargs) as fd:
             yield fd
         log.info("Closing file %s", self._path)
 
-    def copy(self, destination: Union[Path, str]) -> "SecurePath":
+    def move(self, destination: Union[Path, str]) -> "SecurePath":
```
[the remark this answers is not in the document]
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/api/secure_path.py:146]: @sfc-gh-astus `"string"` is a solution, when you return `not-yet-defined` type. There is something like `typing.Self`, but it does not work well with mypy (part of pre-commit)

@sfc-gh-turbaszek Discussed on slack

**Review by sfc-gh-bdufour — APPROVED** [review]
Native app changes LGTM

File: src/snowflake/cli/api/secure_path.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply src/snowflake/cli/api/secure_path.py:146]: 🙇

File: src/snowflake/cli/api/config.py [hunk 2]
```diff
@@ -97,8 +98,8 @@ def get_plugins_config() -> dict:
         return {}
 
 
-def is_default_logs_path(path: Path) -> bool:
-    return path.resolve() == Path(str(_DEFAULT_LOGS_CONFIG["path"])).resolve()
+def is_default_logs_path(path: SecurePath) -> bool:
+    return path.path.resolve() == Path(str(_DEFAULT_LOGS_CONFIG["path"])).resolve()
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/config.py:102]: Hm, this is a bit weird, isn't it? Can we do `resolve` on `path`? Can `SecurePath` behave like a normal `Path`?

File: src/snowflake/cli/plugins/sql/manager.py [hunk 3]
```diff
@@ -26,7 +27,7 @@ def execute(
         if std_in:
             query = sys.stdin.read()
         elif file:
-            query = file.read_text()
+            query = SecurePath(file).read_text(file_size_limit_mb=128)
```
Comment (sfc-gh-pjob) [inline src/snowflake/cli/plugins/sql/manager.py:30]: IMO SQL files shouldn't have a size limit. They can contain DB dumps which can be large.

File: tests/test_utils.py [hunk 4]
```diff
@@ -21,27 +22,30 @@ def test_prepare_app_zip(
...
 def test_prepare_app_zip_if_exception_is_raised_if_no_source(
     temp_directory_for_app_zip,
 ):
     with pytest.raises(FileNotFoundError) as expected_error:
         snowflake.cli.plugins.snowpark.package.utils.prepare_app_zip(
-            Path("/non/existent/path"), temp_directory_for_app_zip
+            SecurePath("/non/existent/path"), SecurePath(temp_directory_for_app_zip)
         )
 
+    print(expected_error.value.__dict__)
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:40]: Please remove `print`

File: src/snowflake/cli/plugins/sql/manager.py (hunk 3 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/plugins/sql/manager.py]: Done

File: tests/test_utils.py (hunk 4 above)
Comment (sfc-gh-pczajka) [reply tests/test_utils.py]: done!

**Review by sfc-gh-pjob — APPROVED** [review]

**Review by sfc-gh-bdufour — APPROVED** [review]
LGTM for native apps

**Review by sfc-gh-pjob — APPROVED** [review]

**Review by sfc-gh-bdufour — APPROVED** [review]

**Pull request merged**